// @ensdomains/ensjs/dist/utils/contents.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.decodeContenthash = decodeContenthash;
exports.validateContent = validateContent;
exports.isValidContenthash = isValidContenthash;
exports.encodeContenthash = encodeContenthash;

var _contentHash = _interopRequireDefault(require("content-hash"));

var _ethers = require("ethers");

var _bs = _interopRequireDefault(require("bs58"));

var supportedCodecs = ['ipns-ns', 'ipfs-ns', 'swarm-ns', 'onion', 'onion3'];
var utils = _ethers.ethers.utils;

function decodeContenthash(encoded) {
  var decoded, protocolType, error;

  if (encoded.error) {
    return {
      protocolType: null,
      decoded: encoded.error
    };
  }

  if (encoded) {
    try {
      decoded = _contentHash["default"].decode(encoded);

      var codec = _contentHash["default"].getCodec(encoded);

      if (codec === 'ipfs-ns') {
        protocolType = 'ipfs';
      } else if (codec === 'ipns-ns') {
        decoded = _bs["default"].decode(decoded).slice(2).toString();
        protocolType = 'ipns';
      } else if (codec === 'swarm-ns') {
        protocolType = 'bzz';
      } else if (codec === 'onion') {
        protocolType = 'onion';
      } else if (codec === 'onion3') {
        protocolType = 'onion3';
      } else {
        decoded = encoded;
      }
    } catch (e) {
      error = e.message;
    }
  }

  return {
    protocolType: protocolType,
    decoded: decoded,
    error: error
  };
}

function validateContent(encoded) {
  return _contentHash["default"].isHashOfType(encoded, _contentHash["default"].Types.ipfs) || _contentHash["default"].isHashOfType(encoded, _contentHash["default"].Types.swarm);
}

function isValidContenthash(encoded) {
  try {
    var codec = _contentHash["default"].getCodec(encoded);

    return utils.isHexString(encoded) && supportedCodecs.includes(codec);
  } catch (e) {
    console.log(e);
  }
}

function encodeContenthash(text) {
  var content, contentType;
  var encoded = false;

  if (!!text) {
    var matched = text.match(/^(ipfs|ipns|bzz|onion|onion3):\/\/(.*)/) || text.match(/\/(ipfs)\/(.*)/) || text.match(/\/(ipns)\/(.*)/);

    if (matched) {
      contentType = matched[1];
      content = matched[2];
    }

    try {
      if (contentType === 'ipfs') {
        if (content.length >= 4) {
          encoded = '0x' + _contentHash["default"].encode('ipfs-ns', content);
        }
      } else if (contentType === 'ipns') {
        var bs58content = _bs["default"].encode(Buffer.concat([Buffer.from([0, content.length]), Buffer.from(content)]));

        encoded = '0x' + _contentHash["default"].encode('ipns-ns', bs58content);
      } else if (contentType === 'bzz') {
        if (content.length >= 4) {
          encoded = '0x' + _contentHash["default"].fromSwarm(content);
        }
      } else if (contentType === 'onion') {
        if (content.length == 16) {
          encoded = '0x' + _contentHash["default"].encode('onion', content);
        }
      } else if (contentType === 'onion3') {
        if (content.length == 56) {
          encoded = '0x' + _contentHash["default"].encode('onion3', content);
        }
      } else {
        console.warn('Unsupported protocol or invalid value', {
          contentType: contentType,
          text: text
        });
      }
    } catch (err) {
      console.warn('Error encoding content hash', {
        text: text,
        encoded: encoded
      }); //throw 'Error encoding content hash'
    }
  }

  return encoded;
}
