// @ensdomains/ensjs/dist/utils/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getEnsStartBlock = getEnsStartBlock;
exports.validateName = validateName;
exports.isLabelValid = isLabelValid;
Object.defineProperty(exports, "isEncodedLabelhash", {
  enumerable: true,
  get: function get() {
    return _labelhash.isEncodedLabelhash;
  }
});
Object.defineProperty(exports, "isDecrypted", {
  enumerable: true,
  get: function get() {
    return _labelhash.isDecrypted;
  }
});
Object.defineProperty(exports, "decodeLabelhash", {
  enumerable: true,
  get: function get() {
    return _labelhash.decodeLabelhash;
  }
});
Object.defineProperty(exports, "encodeLabelhash", {
  enumerable: true,
  get: function get() {
    return _labelhash.encodeLabelhash;
  }
});
Object.defineProperty(exports, "labelhash", {
  enumerable: true,
  get: function get() {
    return _labelhash.labelhash;
  }
});
Object.defineProperty(exports, "encodeContenthash", {
  enumerable: true,
  get: function get() {
    return _contents.encodeContenthash;
  }
});
Object.defineProperty(exports, "decodeContenthash", {
  enumerable: true,
  get: function get() {
    return _contents.decodeContenthash;
  }
});
Object.defineProperty(exports, "isValidContenthash", {
  enumerable: true,
  get: function get() {
    return _contents.isValidContenthash;
  }
});
Object.defineProperty(exports, "namehash", {
  enumerable: true,
  get: function get() {
    return _namehash.namehash;
  }
});
exports.parseSearchTerm = exports.mergeLabels = exports.checkLabels = exports.emptyAddress = exports.uniq = void 0;

var _ethers = require("ethers");

var _labelhash = require("./labelhash");

var _contents = require("./contents");

var _ethEnsNamehash = require("eth-ens-namehash");

var _namehash = require("./namehash");

//import { checkLabelHash } from '../updaters/preImageDB'
var uniq = function uniq(a, param) {
  return a.filter(function (item, pos) {
    return a.map(function (mapItem) {
      return mapItem[param];
    }).indexOf(item[param]) === pos;
  });
};

exports.uniq = uniq;

var checkLabels = function checkLabels() {
  for (var _len = arguments.length, labelHashes = new Array(_len), _key = 0; _key < _len; _key++) {
    labelHashes[_key] = arguments[_key];
  }

  return labelHashes.map(function (hash) {
    return null;
  });
};

exports.checkLabels = checkLabels;

function getEnsStartBlock(networkId) {
  switch (networkId) {
    case 1:
    case '1':
      return 3327417;

    case 3:
    case '3':
      return 25409;

    default:
      return 0;
  }
} // export const checkLabels = (...labelHashes) =>
//   labelHashes.map(labelHash => checkLabelHash(labelHash) || null)


var mergeLabels = function mergeLabels(labels1, labels2) {
  return labels1.map(function (label, index) {
    return label ? label : labels2[index];
  });
};

exports.mergeLabels = mergeLabels;

function validateName(name) {
  var nameArray = name.split('.');
  var hasEmptyLabels = nameArray.filter(function (e) {
    return e.length < 1;
  }).length > 0;
  if (hasEmptyLabels) throw new Error('Domain cannot have empty labels');
  var normalizedArray = nameArray.map(function (label) {
    return (0, _labelhash.isEncodedLabelhash)(label) ? label : (0, _ethEnsNamehash.normalize)(label);
  });

  try {
    return normalizedArray.join('.');
  } catch (e) {
    throw e;
  }
}

function isLabelValid(name) {
  try {
    validateName(name);

    if (name.indexOf('.') === -1) {
      return true;
    }
  } catch (e) {
    console.log(e);
    return false;
  }
}

var parseSearchTerm = function parseSearchTerm(term, validTld) {
  var regex = /[^.]+$/;

  try {
    validateName(term);
  } catch (e) {
    return 'invalid';
  }

  if (term.indexOf('.') !== -1) {
    var termArray = term.split('.');
    var tld = term.match(regex) ? term.match(regex)[0] : '';

    if (validTld) {
      if (tld === 'eth' && termArray[termArray.length - 2].length < 3) {
        return 'short';
      }

      return 'supported';
    }

    return 'unsupported';
  } else if (_ethers.ethers.utils.isAddress(term)) {
    return 'address';
  } else {
    //check if the search term is actually a tld
    if (validTld) {
      return 'tld';
    }

    return 'search';
  }
};

exports.parseSearchTerm = parseSearchTerm;
var emptyAddress = '0x0000000000000000000000000000000000000000';
exports.emptyAddress = emptyAddress;
