// @ensdomains/ensjs/dist/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getENSContract = getENSContract;
exports.getResolverContract = getResolverContract;
exports.getEnsAddress = getEnsAddress;
Object.defineProperty(exports, "namehash", {
  enumerable: true,
  get: function get() {
    return _utils.namehash;
  }
});
Object.defineProperty(exports, "labelhash", {
  enumerable: true,
  get: function get() {
    return _utils.labelhash;
  }
});
exports["default"] = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _ethers = require("ethers");

var _addressEncoder = require("@ensdomains/address-encoder");

var _ENS = require("@ensdomains/ens/build/contracts/ENS.json");

var _Resolver5 = require("@ensdomains/resolver/build/contracts/Resolver.json");

var _ReverseRegistrar = require("@ensdomains/ens/build/contracts/ReverseRegistrar.json");

var _utils = require("./utils");

var _contents = require("./utils/contents");

var Provider = _ethers.ethers.providers.Provider;
var utils = _ethers.ethers.utils;

function getEnsAddress(networkId) {
  if ([1, 3, 4, 5].includes(parseInt(networkId))) {
    return '0x00000000000C2E074eC69A0dFb2997BA6C7d2e1e';
  }
}

function getResolverContract(_ref) {
  var address = _ref.address,
      provider = _ref.provider;
  return new _ethers.ethers.Contract(address, _Resolver5.abi, provider);
}

function getENSContract(_ref2) {
  var address = _ref2.address,
      provider = _ref2.provider;
  return new _ethers.ethers.Contract(address, _ENS.abi, provider);
}

function getReverseRegistrarContract(_ref3) {
  var address = _ref3.address,
      provider = _ref3.provider;
  return new _ethers.ethers.Contract(address, _ReverseRegistrar.abi, provider);
}

function getAddrWithResolver(_x) {
  return _getAddrWithResolver.apply(this, arguments);
}

function _getAddrWithResolver() {
  _getAddrWithResolver = (0, _asyncToGenerator2["default"])(
  /*#__PURE__*/
  _regenerator["default"].mark(function _callee20(_ref4) {
    var name, key, resolverAddr, provider, nh, _Resolver2, _formatsByName$key, coinType, encoder, addr;

    return _regenerator["default"].wrap(function _callee20$(_context20) {
      while (1) {
        switch (_context20.prev = _context20.next) {
          case 0:
            name = _ref4.name, key = _ref4.key, resolverAddr = _ref4.resolverAddr, provider = _ref4.provider;
            nh = (0, _utils.namehash)(name);
            _context20.prev = 2;
            _Resolver2 = getResolverContract({
              address: resolverAddr,
              provider: provider
            });
            _formatsByName$key = _addressEncoder.formatsByName[key], coinType = _formatsByName$key.coinType, encoder = _formatsByName$key.encoder;
            _context20.next = 7;
            return _Resolver2['addr(bytes32,uint256)'](nh, coinType);

          case 7:
            addr = _context20.sent;

            if (!(addr === '0x')) {
              _context20.next = 10;
              break;
            }

            return _context20.abrupt("return", _utils.emptyAddress);

          case 10:
            return _context20.abrupt("return", encoder(Buffer.from(addr.slice(2), 'hex')));

          case 13:
            _context20.prev = 13;
            _context20.t0 = _context20["catch"](2);
            console.log(_context20.t0);
            console.warn('Error getting addr on the resolver contract, are you sure the resolver address is a resolver contract?');
            return _context20.abrupt("return", _utils.emptyAddress);

          case 18:
          case "end":
            return _context20.stop();
        }
      }
    }, _callee20, null, [[2, 13]]);
  }));
  return _getAddrWithResolver.apply(this, arguments);
}

function setAddrWithResolver(_x2) {
  return _setAddrWithResolver.apply(this, arguments);
}

function _setAddrWithResolver() {
  _setAddrWithResolver = (0, _asyncToGenerator2["default"])(
  /*#__PURE__*/
  _regenerator["default"].mark(function _callee21(_ref5) {
    var name, key, address, resolverAddr, signer, nh, Resolver, _formatsByName$key2, decoder, coinType, addressAsBytes;

    return _regenerator["default"].wrap(function _callee21$(_context21) {
      while (1) {
        switch (_context21.prev = _context21.next) {
          case 0:
            name = _ref5.name, key = _ref5.key, address = _ref5.address, resolverAddr = _ref5.resolverAddr, signer = _ref5.signer;
            nh = (0, _utils.namehash)(name);
            Resolver = getResolverContract({
              address: resolverAddr,
              provider: signer
            });
            _formatsByName$key2 = _addressEncoder.formatsByName[key], decoder = _formatsByName$key2.decoder, coinType = _formatsByName$key2.coinType;

            if (!address || address === '') {
              addressAsBytes = Buffer.from('');
            } else {
              addressAsBytes = decoder(address);
            }

            return _context21.abrupt("return", Resolver['setAddr(bytes32,uint256,bytes)'](nh, coinType, addressAsBytes));

          case 6:
          case "end":
            return _context21.stop();
        }
      }
    }, _callee21);
  }));
  return _setAddrWithResolver.apply(this, arguments);
}

function getContentWithResolver(_x3) {
  return _getContentWithResolver.apply(this, arguments);
}

function _getContentWithResolver() {
  _getContentWithResolver = (0, _asyncToGenerator2["default"])(
  /*#__PURE__*/
  _regenerator["default"].mark(function _callee22(_ref6) {
    var name, resolverAddr, provider, nh, _Resolver3, contentHashSignature, isContentHashSupported, _decodeContenthash, protocolType, decoded, error, value, message;

    return _regenerator["default"].wrap(function _callee22$(_context22) {
      while (1) {
        switch (_context22.prev = _context22.next) {
          case 0:
            name = _ref6.name, resolverAddr = _ref6.resolverAddr, provider = _ref6.provider;
            nh = (0, _utils.namehash)(name);

            if (!(parseInt(resolverAddr, 16) === 0)) {
              _context22.next = 4;
              break;
            }

            return _context22.abrupt("return", _utils.emptyAddress);

          case 4:
            _context22.prev = 4;
            _Resolver3 = getResolverContract({
              address: resolverAddr,
              provider: provider
            });
            contentHashSignature = utils.solidityKeccak256(['string'], ['contenthash(bytes32)']).slice(0, 10);
            _context22.next = 9;
            return _Resolver3.supportsInterface(contentHashSignature);

          case 9:
            isContentHashSupported = _context22.sent;

            if (!isContentHashSupported) {
              _context22.next = 25;
              break;
            }

            _context22.t0 = _contents.decodeContenthash;
            _context22.next = 14;
            return _Resolver3.contenthash(nh);

          case 14:
            _context22.t1 = _context22.sent;
            _decodeContenthash = (0, _context22.t0)(_context22.t1);
            protocolType = _decodeContenthash.protocolType;
            decoded = _decodeContenthash.decoded;
            error = _decodeContenthash.error;

            if (!error) {
              _context22.next = 22;
              break;
            }

            console.log('error decoding', error);
            return _context22.abrupt("return", {
              value: _utils.emptyAddress,
              contentType: 'contenthash'
            });

          case 22:
            return _context22.abrupt("return", {
              value: "".concat(protocolType, "://").concat(decoded),
              contentType: 'contenthash'
            });

          case 25:
            _context22.next = 27;
            return _Resolver3.content(nh);

          case 27:
            value = _context22.sent;
            return _context22.abrupt("return", {
              value: value,
              contentType: 'oldcontent'
            });

          case 29:
            _context22.next = 36;
            break;

          case 31:
            _context22.prev = 31;
            _context22.t2 = _context22["catch"](4);
            message = 'Error getting content on the resolver contract, are you sure the resolver address is a resolver contract?';
            console.warn(message, _context22.t2);
            return _context22.abrupt("return", {
              value: message,
              contentType: 'error'
            });

          case 36:
          case "end":
            return _context22.stop();
        }
      }
    }, _callee22, null, [[4, 31]]);
  }));
  return _getContentWithResolver.apply(this, arguments);
}

function setContenthashWithResolver(_x4) {
  return _setContenthashWithResolver.apply(this, arguments);
}

function _setContenthashWithResolver() {
  _setContenthashWithResolver = (0, _asyncToGenerator2["default"])(
  /*#__PURE__*/
  _regenerator["default"].mark(function _callee23(_ref7) {
    var name, content, resolverAddr, signer, encodedContenthash, Resolver;
    return _regenerator["default"].wrap(function _callee23$(_context23) {
      while (1) {
        switch (_context23.prev = _context23.next) {
          case 0:
            name = _ref7.name, content = _ref7.content, resolverAddr = _ref7.resolverAddr, signer = _ref7.signer;
            encodedContenthash = content;

            if (parseInt(content, 16) !== 0) {
              encodedContenthash = (0, _contents.encodeContenthash)(content);
            }

            Resolver = getResolverContract({
              address: resolverAddr,
              provider: signer
            });
            return _context23.abrupt("return", Resolver.setContenthash((0, _utils.namehash)(name), encodedContenthash));

          case 5:
          case "end":
            return _context23.stop();
        }
      }
    }, _callee23);
  }));
  return _setContenthashWithResolver.apply(this, arguments);
}

function getTextWithResolver(_x5) {
  return _getTextWithResolver.apply(this, arguments);
}

function _getTextWithResolver() {
  _getTextWithResolver = (0, _asyncToGenerator2["default"])(
  /*#__PURE__*/
  _regenerator["default"].mark(function _callee24(_ref8) {
    var name, key, resolverAddr, provider, nh, _Resolver4, addr;

    return _regenerator["default"].wrap(function _callee24$(_context24) {
      while (1) {
        switch (_context24.prev = _context24.next) {
          case 0:
            name = _ref8.name, key = _ref8.key, resolverAddr = _ref8.resolverAddr, provider = _ref8.provider;
            nh = (0, _utils.namehash)(name);

            if (!(parseInt(resolverAddr, 16) === 0)) {
              _context24.next = 4;
              break;
            }

            return _context24.abrupt("return", '');

          case 4:
            _context24.prev = 4;
            _Resolver4 = getResolverContract({
              address: resolverAddr,
              provider: provider
            });
            _context24.next = 8;
            return _Resolver4.text(nh, key);

          case 8:
            addr = _context24.sent;
            return _context24.abrupt("return", addr);

          case 12:
            _context24.prev = 12;
            _context24.t0 = _context24["catch"](4);
            console.warn('Error getting text record on the resolver contract, are you sure the resolver address is a resolver contract?');
            return _context24.abrupt("return", '');

          case 16:
          case "end":
            return _context24.stop();
        }
      }
    }, _callee24, null, [[4, 12]]);
  }));
  return _getTextWithResolver.apply(this, arguments);
}

function setTextWithResolver(_x6) {
  return _setTextWithResolver.apply(this, arguments);
}

function _setTextWithResolver() {
  _setTextWithResolver = (0, _asyncToGenerator2["default"])(
  /*#__PURE__*/
  _regenerator["default"].mark(function _callee25(_ref9) {
    var name, key, recordValue, resolverAddr, signer, nh;
    return _regenerator["default"].wrap(function _callee25$(_context25) {
      while (1) {
        switch (_context25.prev = _context25.next) {
          case 0:
            name = _ref9.name, key = _ref9.key, recordValue = _ref9.recordValue, resolverAddr = _ref9.resolverAddr, signer = _ref9.signer;
            nh = (0, _utils.namehash)(name);
            return _context25.abrupt("return", getResolverContract({
              address: resolverAddr,
              provider: signer
            }).setText(nh, key, recordValue));

          case 3:
          case "end":
            return _context25.stop();
        }
      }
    }, _callee25);
  }));
  return _setTextWithResolver.apply(this, arguments);
}

var Resolver =
/*#__PURE__*/
function () {
  //TODO
  function Resolver(_ref10) {
    var address = _ref10.address,
        ens = _ref10.ens;
    (0, _classCallCheck2["default"])(this, Resolver);
    this.address = address;
    this.ens = ens;
  }

  (0, _createClass2["default"])(Resolver, [{
    key: "name",
    value: function name(_name) {
      return new Name({
        name: _name,
        ens: this.ens,
        provider: this.provider,
        signer: this.signer,
        resolver: this.address
      });
    }
  }]);
  return Resolver;
}();

var Name =
/*#__PURE__*/
function () {
  function Name(options) {
    (0, _classCallCheck2["default"])(this, Name);
    var name = options.name,
        ens = options.ens,
        provider = options.provider,
        signer = options.signer,
        nh = options.namehash,
        resolver = options.resolver;

    if (options.namehash) {
      this.namehash = nh;
    }

    this.ens = ens;
    this.ensWithSigner = this.ens.connect(signer);
    this.name = name;
    this.namehash = (0, _utils.namehash)(name);
    this.provider = provider;
    this.signer = signer;
    this.resolver = resolver;
  }

  (0, _createClass2["default"])(Name, [{
    key: "getOwner",
    value: function () {
      var _getOwner = (0, _asyncToGenerator2["default"])(
      /*#__PURE__*/
      _regenerator["default"].mark(function _callee() {
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                return _context.abrupt("return", this.ens.owner(this.namehash));

              case 1:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function getOwner() {
        return _getOwner.apply(this, arguments);
      }

      return getOwner;
    }()
  }, {
    key: "setOwner",
    value: function () {
      var _setOwner = (0, _asyncToGenerator2["default"])(
      /*#__PURE__*/
      _regenerator["default"].mark(function _callee2(address) {
        return _regenerator["default"].wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                if (address) {
                  _context2.next = 2;
                  break;
                }

                throw new Error('No newOwner address provided!');

              case 2:
                return _context2.abrupt("return", this.ensWithSigner.setOwner(this.namehash, address));

              case 3:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function setOwner(_x7) {
        return _setOwner.apply(this, arguments);
      }

      return setOwner;
    }()
  }, {
    key: "getResolver",
    value: function () {
      var _getResolver = (0, _asyncToGenerator2["default"])(
      /*#__PURE__*/
      _regenerator["default"].mark(function _callee3() {
        return _regenerator["default"].wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                return _context3.abrupt("return", this.ens.resolver(this.namehash));

              case 1:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function getResolver() {
        return _getResolver.apply(this, arguments);
      }

      return getResolver;
    }()
  }, {
    key: "setResolver",
    value: function () {
      var _setResolver = (0, _asyncToGenerator2["default"])(
      /*#__PURE__*/
      _regenerator["default"].mark(function _callee4(address) {
        return _regenerator["default"].wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                if (address) {
                  _context4.next = 2;
                  break;
                }

                throw new Error('No resolver address provided!');

              case 2:
                return _context4.abrupt("return", this.ensWithSigner.setResolver(this.namehash, address));

              case 3:
              case "end":
                return _context4.stop();
            }
          }
        }, _callee4, this);
      }));

      function setResolver(_x8) {
        return _setResolver.apply(this, arguments);
      }

      return setResolver;
    }()
  }, {
    key: "getTTL",
    value: function () {
      var _getTTL = (0, _asyncToGenerator2["default"])(
      /*#__PURE__*/
      _regenerator["default"].mark(function _callee5() {
        return _regenerator["default"].wrap(function _callee5$(_context5) {
          while (1) {
            switch (_context5.prev = _context5.next) {
              case 0:
                return _context5.abrupt("return", this.ens.ttl(this.namehash));

              case 1:
              case "end":
                return _context5.stop();
            }
          }
        }, _callee5, this);
      }));

      function getTTL() {
        return _getTTL.apply(this, arguments);
      }

      return getTTL;
    }()
  }, {
    key: "getResolverAddr",
    value: function () {
      var _getResolverAddr = (0, _asyncToGenerator2["default"])(
      /*#__PURE__*/
      _regenerator["default"].mark(function _callee6() {
        return _regenerator["default"].wrap(function _callee6$(_context6) {
          while (1) {
            switch (_context6.prev = _context6.next) {
              case 0:
                if (!this.resolver) {
                  _context6.next = 4;
                  break;
                }

                return _context6.abrupt("return", this.resolver);

              case 4:
                return _context6.abrupt("return", this.getResolver());

              case 5:
              case "end":
                return _context6.stop();
            }
          }
        }, _callee6, this);
      }));

      function getResolverAddr() {
        return _getResolverAddr.apply(this, arguments);
      }

      return getResolverAddr;
    }()
  }, {
    key: "getAddress",
    value: function () {
      var _getAddress = (0, _asyncToGenerator2["default"])(
      /*#__PURE__*/
      _regenerator["default"].mark(function _callee7(coinId) {
        var resolverAddr, Resolver;
        return _regenerator["default"].wrap(function _callee7$(_context7) {
          while (1) {
            switch (_context7.prev = _context7.next) {
              case 0:
                _context7.next = 2;
                return this.getResolverAddr();

              case 2:
                resolverAddr = _context7.sent;

                if (!(parseInt(resolverAddr, 16) === 0)) {
                  _context7.next = 5;
                  break;
                }

                return _context7.abrupt("return", _utils.emptyAddress);

              case 5:
                Resolver = getResolverContract({
                  address: resolverAddr,
                  provider: this.provider
                });

                if (coinId) {
                  _context7.next = 8;
                  break;
                }

                return _context7.abrupt("return", Resolver['addr(bytes32)'](this.namehash));

              case 8:
                return _context7.abrupt("return", getAddrWithResolver({
                  name: this.name,
                  key: coinId,
                  resolverAddr: resolverAddr,
                  provider: this.provider
                }));

              case 9:
              case "end":
                return _context7.stop();
            }
          }
        }, _callee7, this);
      }));

      function getAddress(_x9) {
        return _getAddress.apply(this, arguments);
      }

      return getAddress;
    }()
  }, {
    key: "setAddress",
    value: function () {
      var _setAddress = (0, _asyncToGenerator2["default"])(
      /*#__PURE__*/
      _regenerator["default"].mark(function _callee8(key, address) {
        var resolverAddr;
        return _regenerator["default"].wrap(function _callee8$(_context8) {
          while (1) {
            switch (_context8.prev = _context8.next) {
              case 0:
                if (key) {
                  _context8.next = 2;
                  break;
                }

                throw new Error('No coinId provided');

              case 2:
                if (address) {
                  _context8.next = 4;
                  break;
                }

                throw new Error('No address provided');

              case 4:
                _context8.next = 6;
                return this.getResolverAddr();

              case 6:
                resolverAddr = _context8.sent;
                return _context8.abrupt("return", setAddrWithResolver({
                  name: this.name,
                  key: key,
                  address: address,
                  resolverAddr: resolverAddr,
                  signer: this.signer
                }));

              case 8:
              case "end":
                return _context8.stop();
            }
          }
        }, _callee8, this);
      }));

      function setAddress(_x10, _x11) {
        return _setAddress.apply(this, arguments);
      }

      return setAddress;
    }()
  }, {
    key: "getContent",
    value: function () {
      var _getContent = (0, _asyncToGenerator2["default"])(
      /*#__PURE__*/
      _regenerator["default"].mark(function _callee9() {
        var resolverAddr;
        return _regenerator["default"].wrap(function _callee9$(_context9) {
          while (1) {
            switch (_context9.prev = _context9.next) {
              case 0:
                _context9.next = 2;
                return this.getResolverAddr();

              case 2:
                resolverAddr = _context9.sent;
                return _context9.abrupt("return", getContentWithResolver({
                  name: this.name,
                  resolverAddr: resolverAddr,
                  provider: this.provider
                }));

              case 4:
              case "end":
                return _context9.stop();
            }
          }
        }, _callee9, this);
      }));

      function getContent() {
        return _getContent.apply(this, arguments);
      }

      return getContent;
    }()
  }, {
    key: "setContenthash",
    value: function () {
      var _setContenthash = (0, _asyncToGenerator2["default"])(
      /*#__PURE__*/
      _regenerator["default"].mark(function _callee10(content) {
        var resolverAddr;
        return _regenerator["default"].wrap(function _callee10$(_context10) {
          while (1) {
            switch (_context10.prev = _context10.next) {
              case 0:
                _context10.next = 2;
                return this.getResolverAddr();

              case 2:
                resolverAddr = _context10.sent;
                console.log(content);
                return _context10.abrupt("return", setContenthashWithResolver({
                  name: this.name,
                  content: content,
                  resolverAddr: resolverAddr,
                  signer: this.signer
                }));

              case 5:
              case "end":
                return _context10.stop();
            }
          }
        }, _callee10, this);
      }));

      function setContenthash(_x12) {
        return _setContenthash.apply(this, arguments);
      }

      return setContenthash;
    }()
  }, {
    key: "getText",
    value: function () {
      var _getText = (0, _asyncToGenerator2["default"])(
      /*#__PURE__*/
      _regenerator["default"].mark(function _callee11(key) {
        var resolverAddr;
        return _regenerator["default"].wrap(function _callee11$(_context11) {
          while (1) {
            switch (_context11.prev = _context11.next) {
              case 0:
                _context11.next = 2;
                return this.getResolverAddr();

              case 2:
                resolverAddr = _context11.sent;
                return _context11.abrupt("return", getTextWithResolver({
                  name: this.name,
                  key: key,
                  resolverAddr: resolverAddr,
                  provider: this.provider
                }));

              case 4:
              case "end":
                return _context11.stop();
            }
          }
        }, _callee11, this);
      }));

      function getText(_x13) {
        return _getText.apply(this, arguments);
      }

      return getText;
    }()
  }, {
    key: "setText",
    value: function () {
      var _setText = (0, _asyncToGenerator2["default"])(
      /*#__PURE__*/
      _regenerator["default"].mark(function _callee12(key, recordValue) {
        var resolverAddr;
        return _regenerator["default"].wrap(function _callee12$(_context12) {
          while (1) {
            switch (_context12.prev = _context12.next) {
              case 0:
                _context12.next = 2;
                return this.getResolverAddr();

              case 2:
                resolverAddr = _context12.sent;
                return _context12.abrupt("return", setTextWithResolver({
                  name: this.name,
                  key: key,
                  recordValue: recordValue,
                  resolverAddr: resolverAddr,
                  signer: this.signer
                }));

              case 4:
              case "end":
                return _context12.stop();
            }
          }
        }, _callee12, this);
      }));

      function setText(_x14, _x15) {
        return _setText.apply(this, arguments);
      }

      return setText;
    }()
  }, {
    key: "setSubnodeOwner",
    value: function () {
      var _setSubnodeOwner = (0, _asyncToGenerator2["default"])(
      /*#__PURE__*/
      _regenerator["default"].mark(function _callee13(label, newOwner) {
        var lh;
        return _regenerator["default"].wrap(function _callee13$(_context13) {
          while (1) {
            switch (_context13.prev = _context13.next) {
              case 0:
                lh = (0, _utils.labelhash)(label);
                return _context13.abrupt("return", this.ensWithSigner.setSubnodeOwner(this.namehash, lh, newOwner));

              case 2:
              case "end":
                return _context13.stop();
            }
          }
        }, _callee13, this);
      }));

      function setSubnodeOwner(_x16, _x17) {
        return _setSubnodeOwner.apply(this, arguments);
      }

      return setSubnodeOwner;
    }()
  }, {
    key: "setSubnodeRecord",
    value: function () {
      var _setSubnodeRecord = (0, _asyncToGenerator2["default"])(
      /*#__PURE__*/
      _regenerator["default"].mark(function _callee14(label, newOwner, resolver) {
        var ttl,
            lh,
            _args14 = arguments;
        return _regenerator["default"].wrap(function _callee14$(_context14) {
          while (1) {
            switch (_context14.prev = _context14.next) {
              case 0:
                ttl = _args14.length > 3 && _args14[3] !== undefined ? _args14[3] : 0;
                lh = (0, _utils.labelhash)(label);
                return _context14.abrupt("return", this.ensWithSigner.setSubnodeRecord(this.namehash, lh, newOwner, resolver, ttl));

              case 3:
              case "end":
                return _context14.stop();
            }
          }
        }, _callee14, this);
      }));

      function setSubnodeRecord(_x18, _x19, _x20) {
        return _setSubnodeRecord.apply(this, arguments);
      }

      return setSubnodeRecord;
    }()
  }, {
    key: "createSubdomain",
    value: function () {
      var _createSubdomain = (0, _asyncToGenerator2["default"])(
      /*#__PURE__*/
      _regenerator["default"].mark(function _callee15(label) {
        var resolverPromise, ownerPromise, _ref11, _ref12, resolver, owner;

        return _regenerator["default"].wrap(function _callee15$(_context15) {
          while (1) {
            switch (_context15.prev = _context15.next) {
              case 0:
                resolverPromise = this.getResolver();
                ownerPromise = this.getOwner();
                _context15.next = 4;
                return Promise.all([resolverPromise, ownerPromise]);

              case 4:
                _ref11 = _context15.sent;
                _ref12 = (0, _slicedToArray2["default"])(_ref11, 2);
                resolver = _ref12[0];
                owner = _ref12[1];
                return _context15.abrupt("return", this.setSubnodeRecord(label, owner, resolver));

              case 9:
              case "end":
                return _context15.stop();
            }
          }
        }, _callee15, this);
      }));

      function createSubdomain(_x21) {
        return _createSubdomain.apply(this, arguments);
      }

      return createSubdomain;
    }()
  }, {
    key: "deleteSubdomain",
    value: function () {
      var _deleteSubdomain = (0, _asyncToGenerator2["default"])(
      /*#__PURE__*/
      _regenerator["default"].mark(function _callee16(label) {
        return _regenerator["default"].wrap(function _callee16$(_context16) {
          while (1) {
            switch (_context16.prev = _context16.next) {
              case 0:
                return _context16.abrupt("return", this.setSubnodeRecord(label, _utils.emptyAddress, _utils.emptyAddress));

              case 1:
              case "end":
                return _context16.stop();
            }
          }
        }, _callee16, this);
      }));

      function deleteSubdomain(_x22) {
        return _deleteSubdomain.apply(this, arguments);
      }

      return deleteSubdomain;
    }()
  }]);
  return Name;
}();

var ENS =
/*#__PURE__*/
function () {
  function ENS(options) {
    (0, _classCallCheck2["default"])(this, ENS);
    var networkId = options.networkId,
        provider = options.provider,
        ensAddress = options.ensAddress;
    var ethersProvider;

    if (Provider.isProvider(provider)) {
      //detect ethersProvider
      ethersProvider = provider;
    } else {
      ethersProvider = new _ethers.ethers.providers.Web3Provider(provider);
    }

    this.provider = ethersProvider;
    this.signer = ethersProvider.getSigner();
    this.ens = getENSContract({
      address: ensAddress ? ensAddress : registries[networkId],
      provider: ethersProvider
    });
  }

  (0, _createClass2["default"])(ENS, [{
    key: "name",
    value: function name(_name2) {
      return new Name({
        name: _name2,
        ens: this.ens,
        provider: this.provider,
        signer: this.signer
      });
    }
  }, {
    key: "resolver",
    value: function resolver(address) {
      return new Resolver({
        ens: this.ens,
        provider: this.provider,
        address: address
      });
    }
  }, {
    key: "getName",
    value: function () {
      var _getName = (0, _asyncToGenerator2["default"])(
      /*#__PURE__*/
      _regenerator["default"].mark(function _callee17(address) {
        var reverseNode, resolverAddr;
        return _regenerator["default"].wrap(function _callee17$(_context17) {
          while (1) {
            switch (_context17.prev = _context17.next) {
              case 0:
                reverseNode = "".concat(address.slice(2), ".addr.reverse");
                _context17.next = 3;
                return this.ens.resolver((0, _utils.namehash)(reverseNode));

              case 3:
                resolverAddr = _context17.sent;
                return _context17.abrupt("return", this.getNameWithResolver(address, resolverAddr));

              case 5:
              case "end":
                return _context17.stop();
            }
          }
        }, _callee17, this);
      }));

      function getName(_x23) {
        return _getName.apply(this, arguments);
      }

      return getName;
    }()
  }, {
    key: "getNameWithResolver",
    value: function () {
      var _getNameWithResolver = (0, _asyncToGenerator2["default"])(
      /*#__PURE__*/
      _regenerator["default"].mark(function _callee18(address, resolverAddr) {
        var reverseNode, reverseNamehash, _Resolver, name;

        return _regenerator["default"].wrap(function _callee18$(_context18) {
          while (1) {
            switch (_context18.prev = _context18.next) {
              case 0:
                reverseNode = "".concat(address.slice(2), ".addr.reverse");
                reverseNamehash = (0, _utils.namehash)(reverseNode);

                if (!(parseInt(resolverAddr, 16) === 0)) {
                  _context18.next = 4;
                  break;
                }

                return _context18.abrupt("return", {
                  name: null
                });

              case 4:
                _context18.prev = 4;
                _Resolver = getResolverContract({
                  address: resolverAddr,
                  provider: this.provider
                });
                _context18.next = 8;
                return _Resolver.name(reverseNamehash);

              case 8:
                name = _context18.sent;
                return _context18.abrupt("return", {
                  name: name
                });

              case 12:
                _context18.prev = 12;
                _context18.t0 = _context18["catch"](4);
                console.log("Error getting name for reverse record of ".concat(address), _context18.t0);

              case 15:
              case "end":
                return _context18.stop();
            }
          }
        }, _callee18, this, [[4, 12]]);
      }));

      function getNameWithResolver(_x24, _x25) {
        return _getNameWithResolver.apply(this, arguments);
      }

      return getNameWithResolver;
    }()
  }, {
    key: "setReverseRecord",
    value: function () {
      var _setReverseRecord = (0, _asyncToGenerator2["default"])(
      /*#__PURE__*/
      _regenerator["default"].mark(function _callee19(name, overrides) {
        var reverseRegistrarAddr, reverseRegistrar;
        return _regenerator["default"].wrap(function _callee19$(_context19) {
          while (1) {
            switch (_context19.prev = _context19.next) {
              case 0:
                _context19.next = 2;
                return this.name('addr.reverse').getOwner('addr.reverse');

              case 2:
                reverseRegistrarAddr = _context19.sent;
                reverseRegistrar = getReverseRegistrarContract({
                  address: reverseRegistrarAddr,
                  provider: this.signer
                });
                return _context19.abrupt("return", reverseRegistrar.setName(name));

              case 5:
              case "end":
                return _context19.stop();
            }
          }
        }, _callee19, this);
      }));

      function setReverseRecord(_x26, _x27) {
        return _setReverseRecord.apply(this, arguments);
      }

      return setReverseRecord;
    }()
  }]);
  return ENS;
}();

exports["default"] = ENS;
