// patch-package/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var chalk_1 = __importDefault(require("chalk"));
var process_1 = __importDefault(require("process"));
var minimist_1 = __importDefault(require("minimist"));
var applyPatches_1 = require("./applyPatches");
var getAppRootPath_1 = require("./getAppRootPath");
var makePatch_1 = require("./makePatch");
var makeRegExp_1 = require("./makeRegExp");
var detectPackageManager_1 = require("./detectPackageManager");
var path_1 = require("./path");
var path_2 = require("path");
var slash = require("slash");
var appPath = getAppRootPath_1.getAppRootPath();
var argv = minimist_1.default(process_1.default.argv.slice(2), {
    boolean: [
        "use-yarn",
        "case-sensitive-path-filtering",
        "reverse",
        "help",
        "version",
    ],
    string: ["patch-dir"],
});
var packageNames = argv._;
console.log(chalk_1.default.bold("patch-package"), 
// tslint:disable-next-line:no-var-requires
require(path_1.join(__dirname, "../package.json")).version);
if (argv.version || argv.v) {
    // noop
}
else if (argv.help || argv.h) {
    printHelp();
}
else {
    var patchDir_1 = slash(path_2.normalize((argv["patch-dir"] || "patches") + path_2.sep));
    if (patchDir_1.startsWith("/")) {
        throw new Error("--patch-dir must be a relative path");
    }
    if (packageNames.length) {
        var includePaths_1 = makeRegExp_1.makeRegExp(argv.include, "include", /.*/, argv["case-sensitive-path-filtering"]);
        var excludePaths_1 = makeRegExp_1.makeRegExp(argv.exclude, "exclude", /package\.json$/, argv["case-sensitive-path-filtering"]);
        var packageManager_1 = detectPackageManager_1.detectPackageManager(appPath, argv["use-yarn"] ? "yarn" : null);
        packageNames.forEach(function (packagePathSpecifier) {
            makePatch_1.makePatch({
                packagePathSpecifier: packagePathSpecifier,
                appPath: appPath,
                packageManager: packageManager_1,
                includePaths: includePaths_1,
                excludePaths: excludePaths_1,
                patchDir: patchDir_1,
            });
        });
    }
    else {
        console.log("Applying patches...");
        var reverse = !!argv["reverse"];
        applyPatches_1.applyPatchesForApp({ appPath: appPath, reverse: reverse, patchDir: patchDir_1 });
    }
}
function printHelp() {
    console.log("\nUsage:\n\n  1. Patching packages\n  ====================\n\n    " + chalk_1.default.bold("patch-package") + "\n\n  Without arguments, the " + chalk_1.default.bold("patch-package") + " command will attempt to find and apply\n  patch files to your project. It looks for files named like\n\n     ./patches/<package-name>+<version>.patch\n\n  2. Creating patch files\n  =======================\n\n    " + chalk_1.default.bold("patch-package") + " <package-name>" + chalk_1.default.italic("[ <package-name>]") + "\n\n  When given package names as arguments, patch-package will create patch files\n  based on any changes you've made to the versions installed by yarn/npm.\n\n  Options:\n\n     " + chalk_1.default.bold("--use-yarn") + "\n\n         By default, patch-package checks whether you use npm or yarn based on\n         which lockfile you have. If you have both, it uses npm by default.\n         Set this option to override that default and always use yarn.\n\n     " + chalk_1.default.bold("--exclude <regexp>") + "\n\n         Ignore paths matching the regexp when creating patch files.\n         Paths are relative to the root dir of the package to be patched.\n\n         Default: 'package\\.json$'\n\n     " + chalk_1.default.bold("--include <regexp>") + "\n\n         Only consider paths matching the regexp when creating patch files.\n         Paths are relative to the root dir of the package to be patched.\n\n         Default '.*'\n\n     " + chalk_1.default.bold("--case-sensitive-path-filtering") + "\n\n         Make regexps used in --include or --exclude filters case-sensitive.\n");
}
